pkg/deviceplugins/device_manager.go

```diff
@@ -0,0 +1,490 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2023 SUSE, LLC.
+ *
+ */
+package deviceplugins
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+
+	"kubevirt.io/client-go/log"
+
+	"kubevirt.io/kubevirt/pkg/util"
+	pluginapi "kubevirt.io/kubevirt/pkg/virt-handler/device-manager/deviceplugin/v1beta1"
+)
+
+const (
+	vfioDevicePath    = "/dev/vfio/"
+	vfioMount         = "/dev/vfio/vfio"
+	pciBasePath       = "/sys/bus/pci/devices"
+	connectionTimeout = 120 * time.Second // Google gRPC default timeout
+	PCIResourcePrefix = "PCI_RESOURCE"
+)
+
+type PCIDevice struct {
+	pciID      string
+	driver     string
+	pciAddress string
+	iommuGroup string
+	numaNode   int
+}
+
+type PCIDevicePlugin struct {
+	pcidevs       []*PCIDevice
+	devs          []*pluginapi.Device
+	server        *grpc.Server
+	socketPath    string
+	stop          <-chan struct{}
+	health        chan deviceHealth
+	devicePath    string
+	resourceName  string
+	done          chan struct{}
+	deviceRoot    string
+	iommuToPCIMap map[string]string
+	initialized   bool
+	lock          *sync.Mutex
+	deregistered  chan struct{}
+	starter       *DeviceStarter
+}
+
+type DeviceStarter struct {
+	started  bool
+	stopChan chan struct{}
+	backoff  []time.Duration
+}
+
+func (dp *PCIDevicePlugin) GetPCIDevices() []*PCIDevice {
+	return dp.pcidevs
+}
+
+// Not adding more data to the struct, it's big enough already
+func (dp *PCIDevicePlugin) GetCount() int {
+	var count int
+	for _, dev := range dp.devs {
+		if dev.Health == pluginapi.Healthy {
+			count += 1
+		}
+	}
+	return count
+}
+
+func (d *PCIDevice) GetID() string {
+	return d.pciID
+}
+
+func NewPCIDevicePlugin(pciDevices []*PCIDevice, resourceName string) *PCIDevicePlugin {
+	serverSock := SocketPath(strings.Replace(resourceName, "/", "-", -1))
+	iommuToPCIMap := make(map[string]string)
+
+	initHandler()
+
+	devs := constructDPIdevices(pciDevices, iommuToPCIMap)
+	dpi := &PCIDevicePlugin{
+		pcidevs:       pciDevices,
+		devs:          devs,
+		socketPath:    serverSock,
+		resourceName:  resourceName,
+		devicePath:    vfioDevicePath,
+		deviceRoot:    util.HostRootMount,
+		iommuToPCIMap: iommuToPCIMap,
+		health:        make(chan deviceHealth),
+		initialized:   false,
+		lock:          &sync.Mutex{},
+		starter: &DeviceStarter{
+			started:  false,
+			stopChan: make(chan struct{}),
+			backoff:  defaultBackoffTime,
+		},
+	}
+	return dpi
+}
+
+func constructDPIdevices(pciDevices []*PCIDevice, iommuToPCIMap map[string]string) (devs []*pluginapi.Device) {
+	for _, pciDevice := range pciDevices {
+		iommuToPCIMap[pciDevice.iommuGroup] = pciDevice.pciAddress
```

Comment:
```suggestion
		iommuToPCIMap[pciDevice.pciAddress] = pciDevice.iommuGroup
```
swapped struct of map to make it easy to lookup iommuGroup from the device address